File: packages/python/src/rut_cl/check_digit.py
from __future__ import annotations

from rut_cl.clean import clean


def check_digit(input: object) -> str:
    """Compute the modulo-11 check digit for a RUT body (digits only, no DV)."""
    cleaned = clean(input)

    if cleaned == "" or not cleaned.isdigit():
        raise ValueError(f'"{input}" as RUT is invalid')

    digits = [int(char) for char in cleaned]
    total = sum(
        digit * ((index % 6) + 2) for index, digit in enumerate(reversed(digits))
    )
    digit = 11 - (total % 11)

    if digit == 10:
        return "K"
    if digit == 11:
        return "0"
    return str(digit)

File: packages/js/src/check-digit.ts
/**
 * Compute the modulo-11 check digit for a RUT body (digits only, no DV).
 */
export function checkDigit(body: string): string {
  const digits = Array.from(body, Number)

  if (digits.length === 0 || digits.includes(Number.NaN)) {
    throw new Error(`"${String(body)}" as RUT is invalid`)
  }

  const sum = [...digits].reverse().reduce(
    (accumulator, currentValue, index) =>
      accumulator + currentValue * ((index % 6) + 2),
    0,
  )

  const digit = 11 - (sum % 11)

  if (digit === 10) {
    return 'K'
  }

  if (digit === 11) {
    return '0'
  }

  return String(digit)
}
